### PR TITLE
Flow control

### PR DESCRIPTION
**Pull request by zenhack**
Still much to do here (including docs), but I have started working on support for per-object flow control. The plan is to add a `SetFlowLimiter` method to `Client`. Possibly after this is merged we can also add a `FlowLimiter` that is adaptive, based on BBR or the like, rather than just using a fixed window.

### CODE REVIEW

**Comment by zenhack** [comment]
Alright, this is finished & tested, PTAL.

As a separate PR, at some point, we should provide an adaptive implementation of FlowLimiter, but this gets the machinery in place and at least lets you do flow control if you want to tell the library what the right size is.

**Comment by zenhack** [comment]
Ok, looks like I need to fuss with the test a bit more; it passes on my machine, so perhaps I'm just not providing enough slack for github CI. I'll have to play with it so that it's loose enough to not tigger false positives, but not so loose that it doesn't catch real bugs.

Annoyingly, this kind of thing is sortof inherently racy, so it's a bit of an art figuring out how much slack is enough...

**Comment by zenhack** [comment]
(In the meantime, this is still worth reviewing)

File: rpc/flow_test.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	"github.com/stretchr/testify/assert"
+
+	"capnproto.org/go/capnp/v3"
+	"capnproto.org/go/capnp/v3/flowcontrol"
+	"capnproto.org/go/capnp/v3/rpc/internal/testcapnp"
+	"capnproto.org/go/capnp/v3/server"
+	rpccp "capnproto.org/go/capnp/v3/std/capnp/rpc"
+)
+
+// Transport, which is a wrapper around another transport, and measures the total size of
```
Comment (lthibault) [inline rpc/flow_test.go:21]: `measuringTransport`

**Review by lthibault — APPROVED** [review]

**Comment by zenhack** [comment]
@lthibault can you check out this branch and tell me what the new tests do on your end? Also open to specific ideas wrt. how to make the failing test more robust.

One thing I noticed when poking around: allocating a slot for the outstanding call on the server is a linear time operation. In the test I cranked `MaxConcurrentCalls` *way* up to avoid a situation where we weren't hitting the flow limit because the receive loop was blocking on a full server queue. But, this could result in actually spending a lot of CPU time in that allocation loop, and I wonder if in an environment with less available CPU that could still be a bottleneck. The CI run for 3f3cc1c reported notably lower flow on Go 1.15 than Go 1.17, which, assuming the compiler gets better over time, supports the idea that this may be a CPU bound problem.

Because of that, unless you have any great insights, or running it on your end leads to some discovery, I wonder if it wouldn't make sense to table the test until we've gotten rid of that queue, since we're planning on doing that anyway, and then pick it up afterwards. I think I don't want to leave the PR hanging though, since it touches some bits of the code I want to work on for #200; I suppose we could leave it in with a `t.Skip(...)` for now...

**Comment by lthibault** [comment]
@zenhack 

Seems to be working fine for me.  Though did you perhaps remove the offending test?  `TestFixedFlowLimit` is the test that fails, but I only see `TestFixed` in the current tip of your branch (presumably, this is just a renaming).

>this could result in actually spending a lot of CPU time in that allocation loop

This would be `fixedLimiter.pumpQueue`, yes?

Perhaps it would be more performant to use [golang.org/x/sync/semaphore](https://pkg.go.dev/golang.org/x/sync/semaphore), here?  This avoids the busy loop, which is presumably what's burning up CPU cycles.

>I wonder if it wouldn't make sense to table the test until we've gotten rid of that queue
> I think I don't want to leave the PR hanging though

Yes, this sounds reasonable to me 👍 

```
$ go test -v ./flowcontrol/...
=== RUN   TestFixed
--- PASS: TestFixed (0.02s)
PASS
ok      capnproto.org/go/capnp/v3/flowcontrol   0.191s
``` 

```
$ go test -v -race ./flowcontrol/...
=== RUN   TestFixed
--- PASS: TestFixed (0.02s)
PASS
ok      capnproto.org/go/capnp/v3/flowcontrol   0.200s
```

**Comment by zenhack** [comment]
Quoting Louis Thibault (2021-12-19 14:57:18)

>    Seems to be working fine for me. Though did you perhaps remove the
>    offending test? TestFixedFlowLimit is the test that fails, but I only
>    see TestFixed in the current tip of your branch (presumably, this is
>    just a renaming).

No, these are two different tests and they should both be there. The
failing test is in rpc/, not flowcontrol/.

>    This would be fixedLimiter.pumpQueue, yes?

No, I'm talking about Server.nextID in server/server.go, which is pre-existing
code.

**Comment by zenhack** [comment]
Disabled the test, per discussion on matrix.

**Comment by zenhack** [comment]
@lthibault, are you comfortable merging this?

**Review by lthibault — APPROVED** [review]

**Pull request merged**